_posts/2015-12-16-adding-csrf-protection-to-your-rails-backbone-app.markdown

```diff
@@ -0,0 +1,58 @@
+---
+layout: post
+title: "Adding Csrf-Protection to your Rails-Backbone App"
+date: 2015-12-16 03:55:00
+categories: ["rails", "backbone", "security"]
+author: "schmierkov"
+---
+
+When integrating Backbone.js in your Rails App, you might face the problem of the inability to verify the CSRF-Token.
+
+The CSRF Protection secures your app with a token to make sure, that the person who is interacting with it is you. So you should not turn it off, except you know what you are doing. For more informations on this Topic, check out the [Rails Security Guide](http://guides.rubyonrails.org/security.html#cross-site-request-forgery-csrf).
+
+This problem occurs as soon as you are trying to send form data, without the csrf-token provided by Rails.
+
+```ruby
+Started POST "/products" for 127.0.0.1 at 2015-12-16 10:06:05 -0300
+Processing by ProductsController#create as JSON
+  Parameters: {"product"=>{"name"=>"foo"}}
+WARNING: Can't verify CSRF token authenticity
+...
+Completed 302 Found in 7.6ms (ActiveRecord: 0.8ms)
+```
+
+After this request, rails will terminate your session and you have to login again.
+
+This problem is caused by your Backbone.js application, which is sending the data directly to the backend without providing the CSRF-Token.
+
+To solve this problem you need to add the token to your Backbone request. One of the simplest solutions I came across is the following by [Anton Shuvalov](https://github.com/shuvalov-anton/backbone-rails-sync).
+
+```javascript
+// https://github.com/shuvalov-anton/backbone-rails-sync
+Backbone._sync = Backbone.sync;
+Backbone.sync = function(method, model, options) {
+  if (!options.noCSRF) {
+    var beforeSend = options.beforeSend;
+
+    // Set X-CSRF-Token HTTP header
+    options.beforeSend = function(xhr) {
+      var token = $('meta[name="csrf-token"]').attr('content');
+      if (token) { xhr.setRequestHeader('X-CSRF-Token', token); }
+      if (beforeSend) { return beforeSend.apply(this, arguments); }
+    };
+  }
+  return Backbone._sync(method, model, options);
+};
+```
+
+It grabs the csrf-token provided in the meta tags of your rails application and sets this as request header field `X-CSRF-Token`.
```

Comment:
@schmierkov I'd stick either to `CSRF-token` or `csrf-token`. Also capitalize Rails where it's not capitalized, there's currently a mix of both of these.